### ISSUE TITLE
Human bots deals way too much damage to teammates

### ISSUE THREAD

**Issue by sweet235** [Unvanquished/Unvanquished#2775]
To see the problem, load map-operation-dretch for instance. Build an armoury next to the telenodes. Then do this:

```/g_freeFundPeriod 1; g_friendlyFireHumanMultiplier 100; g_bot_lcannon off; g_bot_prifle off; g_bot_flamer off; g_bot_chain off; g_bot_mdriver off; bot fill 10 h; bot skill 9 h```

This will add 10 human bots with skill 9 who can instagib teammates. They will start to destroy eggs. Soon, they will all have lasguns. The lasgun is hitscan and perfectly accurate. Yet we can observe this:

```
●[bot] Exior K'Mool was lasered by TEAMMATE ●[bot] Edward Pickman Derby 
●[bot] Abner Exekiel Hoag was lasered by TEAMMATE ●[bot] Zadok Allen 
●[bot] Malo was lasered by TEAMMATE ●[bot] Zadok Allen 
●[bot] Crom-Ya was lasered by TEAMMATE ●[bot] Ambrose Dexter 
●[bot] Bran Mak Morn was lasered by TEAMMATE ●[bot] Gustaf Johansen 
...
```

This is most likely not because of the bots' aim being too slow, as skill 9 bots are pretty much perfect aimbots. Indeed, reducing `g_bot_humanAimDelay` does not change this situation.

Why does this happen so often?

**Comment by sweet235** [comment]
I should add that our bots' friendly fire problem is severe enough to force me to prevent any bot damage to users on their team on my server.

**Comment by illwieckz** [comment]
Implementing smaller friendly hitbox would help!

**Comment by VReaperV** [comment]
But also, I've seen bots turn around 180 degrees and shoot a teammate in the face for no reason, with no enemies and the edge of the map behind them.

**Comment by slipher** [comment]
Yeah I was working on a mod recently and found out that in BotActionFight there is no verification whatsoever that the bot is actually facing toward an enemy. It just starts shooting as long as the enemy is "visible" (which considers only the world) and within weapon range.

**Comment by sweet235** [comment]
> I think this is because bots ignore any players/structures when they're attacking their target

> ... which considers only whether a trace between the two entities is obstructed by the world

It seems to me this is both true and false to some degree. I tried to step in between a bot shooting an egg. If I step at the right place, the bot will stop to fire and start to move until it can see it again. You can easily verify this in the test setup I proposed above.

I think the trace in `BotTargetIsVisible` works sometimes, but fails at other times. For reference:
https://github.com/Unvanquished/Unvanquished/blob/master/src/sgame/sg_bot_util.cpp#L1440

**Comment by sweet235** [comment]
> Implementing smaller friendly hitbox would help!

I do not think so. It would reduce the likelihood a little, but once the bots form a crowd, this will happen.

**Comment by sweet235** [comment]
> I've seen bots try to shoot at structures with some other buildings in-between, sometimes dead buildings

I think dead buildings are completely ignored, and living teammates only in some cases.

**Comment by sweet235** [comment]
Well, I might have been wrong above. There are actually 2 conditions to be met to make a bot fire:

- BotTargetIsVisible must be true
- BotTargetInAttackRange must be true

Turns out that both do traces.

**Comment by sweet235** [comment]
> so that could explain some cases

It does not explain what I reported. I verified this by trying 0.54 and 0.53, just to be sure.

**Comment by slipher** [comment]
> Yeah I was working on a mod recently and found out that in BotActionFight there is no verification whatsoever that the bot is actually facing toward an enemy.

What I said here is still correct. All the traces are done along the line connecting the bot and the enemy. There is no trace done using the bot's actual forward direction.

**Comment by sweet235** [comment]
> All the traces are done along the line connecting the bot and the enemy. There is no trace done using the bot's actual forward direction.

After all movement/oritientation has stopped, and nobody moves: should there be a difference between the two?

**Comment by slipher** [comment]
> After all movement/oritientation has stopped, and nobody moves: should there be a difference between the two?

Obviously not if there are no bugs, but bots do not wait until orientation has completed to begin firing.

**Comment by bmorel** [comment]
> There is no trace done using the bot's actual forward direction

https://github.com/Unvanquished/Unvanquished/blob/master/src/sgame/sg_bot_util.cpp#L1988-L1993

One of the obvious problems here is: bots do trace in the forward direction, but they don't check the missiles' bbox if they're using one. Nor do they check for the "wide trace" one would need to check for melee attacks.

But yes, indeed, they don't use their forward vector to trace, but rather whatever is returned by "BotGetIdealAimLocation". Which does not use the real location for head, notably, but a hard coded value that simulates it (and will break the 1st time anything in there is changed).

I would not be surprised if the MASK_SHOT does not really do what is expected, neither. The trace masks are rather obscure, and I always have to dig daemon's code to have them make some (and really, I mean it: *some*) sense to me.

**Comment by slipher** [comment]
I mean, there is no trace in the actual forward direction until the decision to fire has already been taken :) Only in the case of luci, apparently, may the human bot decide not to fire after calling the "fire weapon" function.
